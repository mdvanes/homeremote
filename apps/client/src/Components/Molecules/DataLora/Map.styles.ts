import { makeStyles } from "@mui/material";

const useStyles = makeStyles(() => ({
    page: {
        height: "calc(100vh - 80px)",
        overflow: "clip",
    },
    card: {
        aspectRatio: "16/9",
        marginTop: 10,
        overflow: "clip",
    },
    map: {
        height: "100%",
        position: "relative",
        "& .leaflet-container": {
            height: "100%",
        },
        "& .custom-controls": {
            position: "absolute",
            top: 0,
            right: 0,
            zIndex: 1000,
            padding: "1rem",
            "& button": {
                textTransform: "uppercase",
                padding: "0.5rem",
                marginLeft: "0.5rem",
                fontWeight: "bold",
            },
        },
    },
}));

export default useStyles;
