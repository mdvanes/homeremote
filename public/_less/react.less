// for /react test

// Toggle
.btn-icon {
    font-size: 350%;
}

// Button-Set
.btn-group-justified {
    > a,
    > span {
        font-size: 230%;
    }
}

.margin-top {
    margin-top: 1em;
}

.log {
    color: #fff;
    height: 300px;
}