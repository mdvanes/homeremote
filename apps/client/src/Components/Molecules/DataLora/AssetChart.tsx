import { TrackerItem } from "@homeremote/types";
import { FC } from "react";
import { Line, LineChart, Tooltip, XAxis, YAxis } from "recharts";
import useStyles from "./Map.styles";

interface Props {
    coords: TrackerItem[][];
    setActiveMarkerTimestamp: (timestamp: string) => void;
}

export const AssetsChart: FC<Props> = ({
    coords,
    setActiveMarkerTimestamp,
}) => {
    const { classes } = useStyles();

    const asset1 = coords[0]; // Traccar batteryLevel is not in an attribute but in a seperate sensor entity. Even then, it only stores the last value. It could be solved to create a template sensor in HASS that combines the loc and battery.
    const data = asset1
        ? asset1.map((item) => ({
              battery: item.batteryLevel, // ? parseInt(item.battery, 10) : -1,
              time: item.time,
              // temperature: item.temperature, // ? parseInt(item.temperature, 10) : -1,
          }))
        : [];

    return (
        <div className={classes.chart}>
            <LineChart width={500} height={150} data={data}>
                <Line
                    yAxisId="left"
                    type="monotone"
                    dataKey="battery"
                    stroke="#8884d8"
                    dot={false}
                />
                <YAxis
                    yAxisId="left"
                    orientation="left"
                    label="%"
                    width={75}
                    domain={[0, 100]}
                />
                <XAxis
                    dataKey="time"
                    tickFormatter={(value) =>
                        new Date(value.toString()).toLocaleString("en-uk", {
                            hour: "2-digit",
                            minute: "2-digit",
                        })
                    }
                />
                <Tooltip
                    labelFormatter={(value) =>
                        new Date(value.toString()).toLocaleString("en-uk")
                    }
                    formatter={(value, name, entry) => {
                        if (entry.payload.time) {
                            setActiveMarkerTimestamp(entry.payload?.time ?? "");
                        }

                        if (name === "battery") {
                            return `${value}%`;
                        }
                        if (name === "temperature") {
                            return `${value}Â°C`;
                        }
                        return value;
                    }}
                />
            </LineChart>
        </div>
    );
};
