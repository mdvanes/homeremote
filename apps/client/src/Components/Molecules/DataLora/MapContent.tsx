import { polygon } from "leaflet";
import { FC, useCallback, useEffect, useState } from "react";
import { TileLayer, Marker, Popup, Polyline, useMap } from "react-leaflet";
import { TrackerItem } from "@homeremote/types";

// const DEFAULT_CENTER: LatLngTuple = [52, 5.1];

// const TILES_LAYER_DEFAULT =
//     "https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png";
// const TILES_LAYER_BW = "https://tiles.wmflabs.org/bw-mapnik/{z}/{x}/{y}.png";
// Source: https://leaflet-extras.github.io/leaflet-providers/preview/
const TILES_LAYER_DARK =
    "https://{s}.basemaps.cartocdn.com/dark_all/{z}/{x}/{y}{r}.png";

interface Props {
    coords: TrackerItem[];
}

const MapContent: FC<Props> = ({ coords }) => {
    const [marker, setMarker] = useState<TrackerItem | null>(null);
    const map = useMap();

    const updateBoundsAndMarker = useCallback(() => {
        if (coords.length > 0) {
            const newPoly = polygon(coords.map(({ loc }) => loc));
            map.fitBounds(newPoly.getBounds());
            const last = coords[coords.length - 1];
            setMarker(last);
        }
    }, [coords, map]);

    useEffect(() => {
        updateBoundsAndMarker();
    }, [coords, updateBoundsAndMarker]);

    return (
        <>
            <TileLayer
                attribution='&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors'
                url={TILES_LAYER_DARK}
            />
            <Polyline positions={coords.map((coord) => coord.loc)} />
            {marker && (
                <Marker position={marker.loc}>
                    <Popup>
                        {new Date(marker.time).toLocaleString("nl-nl")}
                    </Popup>
                </Marker>
            )}
        </>
    );
};

export default MapContent;
