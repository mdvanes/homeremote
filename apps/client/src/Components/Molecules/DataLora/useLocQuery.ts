import { TrackerItem, TrackerQueryType } from "@homeremote/types";
import { useState } from "react";
import { useGetCoordsQuery } from "../../../Services/dataloraApi";

const UPDATE_INTERVAL_MS = 1000 * 60 * 60; // 1000 ms / 60 seconds / 60 minutes = 1x per hour

interface LocQuery {
    coords: TrackerItem[];
    update: () => Promise<void>;
    toggleQueryType: () => void;
    isLoading: boolean;
    queryType: TrackerQueryType;
}

export const useLocQuery = (): LocQuery => {
    const [queryType, setQueryType] = useState<TrackerQueryType>("24h");
    const {
        data: coords,
        isLoading,
        isFetching,
        refetch,
    } = useGetCoordsQuery(
        {
            type: queryType,
        },
        { pollingInterval: UPDATE_INTERVAL_MS, refetchOnMountOrArgChange: true }
    );

    const update = async (): Promise<void> => {
        refetch();
    };

    const toggleQueryType = () => {
        const newType = queryType === "all" ? "24h" : "all";
        setQueryType(newType);
    };

    return {
        coords: coords || [],
        update,
        isLoading: isLoading || isFetching,
        toggleQueryType,
        queryType,
    };
};
