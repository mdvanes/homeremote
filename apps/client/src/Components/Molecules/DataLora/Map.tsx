import { FC } from "react";
import { MapContainer } from "react-leaflet";
import { AssetsChart } from "./AssetChart";
import useStyles from "./Map.styles";
import MapContent from "./MapContent";
import { useLocQuery } from "./useLocQuery";

interface Props {
    showChart?: boolean;
}

const Map: FC<Props> = ({ showChart = false }) => {
    const { classes } = useStyles();
    const { coords, update, isLoading, toggleQueryType, queryType } =
        useLocQuery();

    return (
        <div className={classes.map}>
            <MapContainer zoom={20}>
                <MapContent coords={coords} />
            </MapContainer>
            <div className="custom-controls">
                <button onClick={() => update()} disabled={isLoading}>
                    update
                </button>
                <button onClick={toggleQueryType} disabled={isLoading}>
                    {queryType}
                </button>
            </div>
            {showChart && coords && <AssetsChart coords={coords} />}
        </div>
    );
};

export default Map;
