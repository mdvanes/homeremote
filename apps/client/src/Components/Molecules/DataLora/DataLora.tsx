import { Paper } from "@mui/material";
import { FC } from "react";
import Map from "./Map";
import useStyles from "./Map.styles";

const DataLora: FC = () => {
    const { classes } = useStyles();
    return (
        <Paper className={classes.card}>
            <Map />
        </Paper>
    );
};

export default DataLora;
