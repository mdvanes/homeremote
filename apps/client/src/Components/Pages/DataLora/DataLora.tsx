import { FC } from "react";
import { Paper } from "@mui/material";
import Map from "../../Molecules/DataLora/Map";
import useStyles from "../../Molecules/DataLora/Map.styles";

const DataLoraPage: FC = () => {
    const { classes } = useStyles();
    return (
        <Paper className={classes.page}>
            <Map />
        </Paper>
    );
};

export default DataLoraPage;
