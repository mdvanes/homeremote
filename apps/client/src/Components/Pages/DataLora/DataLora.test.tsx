import React from "react";
import { render } from "@testing-library/react";
import DataLora from "./DataLora";

jest.mock("../../Molecules/DataLora/Map", () => "mock-map");

describe("DataLora page", () => {
    it("contains the Map component", () => {
        const { baseElement } = render(<DataLora />);
        expect(baseElement).toMatchSnapshot();
    });
});
