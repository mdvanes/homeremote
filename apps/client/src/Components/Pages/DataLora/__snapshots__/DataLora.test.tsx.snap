// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataLora page contains the Map component 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root makeStyles-page MuiPaper-elevation1 MuiPaper-rounded"
    >
      <mock-map />
    </div>
  </div>
</body>
`;
